tidy3d/plugins/waveguide/rectangular_dielectric.py

```diff
@@ -0,0 +1,560 @@
+
+    @pydantic.validator("gap", always=True)
+    def _validate_gaps(cls, val, values):
+        if val.size == 1 and values["core_width"].size != 2:
+            return numpy.array([val[0]] * (values["core_width"].size - 1))
+        if val.size != values["core_width"].size - 1:
+            raise ValidationError("Number of gaps must be 1 less than number of core widths.")
+        return val
+
+    @pydantic.root_validator
+    def _ensure_consistency(cls, values):
+        if values["sidewall_thickness"] > 0 and values["sidewall_medium"] is None:
+            raise ValidationError(
+                "Sidewall medium must be provided when sidewall thickness is greater than 0."
+            )
+
+        if values["sidewall_thickness"] == 0 and values["sidewall_medium"] is not None:
+            log.warning("Sidewall medium not used because sidewall thickness is zero.")
+
+        if values["surface_thickness"] > 0 and values["surface_medium"] is None:
+            raise ValidationError(
+                "Surface medium must be provided when surface thickness is greater than 0."
+            )
+
+        if values["surface_thickness"] == 0 and values["surface_medium"] is not None:
+            log.warning("Surface medium not used because surface thickness is zero.")
+
+        if values["propagation_axis"] == values["normal_axis"]:
+            raise ValidationError("Propagation and normal axes must be different.")
+
+        return values
+
+    @cached_property
+    def lateral_axis(self):
+        """Lateral direction axis"""
+        return 3 - self.propagation_axis - self.normal_axis
+
+    def _transform(self, lateral_coordinate, normal_coordinate, propagation_coordinate):
+        """Swap the model coordinates to desired axes and translate to origin"""
+        result = list(self.origin)
+        result[self.lateral_axis] += lateral_coordinate
+        result[self.propagation_axis] += propagation_coordinate
+        result[self.normal_axis] += normal_coordinate
+        return result
+
+    def _transform_in_plane(self, lateral_coordinate, propagation_coordinate):
+        """Swap the model coordinates to desired axes in the substrate plane"""
+        result = self._transform(lateral_coordinate, 0, propagation_coordinate)
+        result.pop(self.normal_axis)
+        return result
+
+    @cached_property
+    def height(self):
+        """Domain height (size in the normal direction)"""
+        return self.box_thickness + self.core_thickness + self.clad_thickness
+
+    @cached_property
+    def width(self):
+        """Domain width (size in the lateral direction)"""
+        w = self.core_width.sum() + self.gap.sum() + 2 * self.side_margin
+        if self.sidewall_angle > 0:
+            w += 2 * self.core_thickness * numpy.tan(self.sidewall_angle)
+        return w
+
+    # pylint:disable=too-many-locals,too-many-statements,too-many-branches
+    @cached_property
```

Comment:
this method should be broken up I think.

Reply:
This function is long because it creates the structures and override structures for the mode solver for any setting the user might pass, so it handles any cases the class supports. The thing is that each part of the code depends on the variables calculated before it, so, if we break it up into smaller functions, we'd have to bass a bunch of variables around and that list would just keep increasing along the calculation. Furthermore, I don't think there's any piece of code here that could be reused somewhere else. Breaking this up would mean calling a series of different functions in a precise order passing a lot of arguments that would lose their meaning in the context of an isolated function, so the result would be an increase in complexity, in my opinion.

Reply:
I'm not sure I agree actually. It seems to me that the function has a series of pretty logically distinct operations, which could be captured into their own methods. At the very least, probably setting up the grid spec, and creating the structures is the first step (the method name even mentions doing two things). But I think it would help to be refactored with more granularity.  My worry is that in its current state, this method would be a bit difficult for another developer to come in and extend / edit. It also doesn't lend itself well to testing each step of the process. Do you have a good feeling for how much complexity refactoring this method would introduce? To avoid passing too many arguments, I think adding some functions defined locally to `_structures_and_gridspec()` could be an option.

Reply:
Also, what are your ideas for the future of this plugin? do you plan to extend it in any way? or use it in other plugins?